### PR TITLE
Clean code, mainly scoping variables

### PR DESCRIPTION
**Pull request by alkino**

### CODE REVIEW

File: coreneuron/gpu/nrn_acc_manager.cpp [hunk 1]
```diff
@@ -856,10 +835,10 @@ void nrn_newtonspace_copyto_device(NewtonSpace* ns) {
     pd = (double*)acc_copyin(ns->rowmax, n * sizeof(double));
     acc_memcpy_to_device(&(d_ns->rowmax), &pd, sizeof(double*));
 
-    pint = (int*)acc_copyin(ns->perm, n * sizeof(int));
+    int* pint = (int*)acc_copyin(ns->perm, n * sizeof(int));
```
Comment (tristan0x) [inline coreneuron/gpu/nrn_acc_manager.cpp]: I usually avoid duplicating the type

```suggestion
    auto pint = static_cast<int*>(acc_copyin(ns->perm, n * sizeof(int)));
```

File: coreneuron/gpu/nrn_acc_manager.cpp [hunk 2]
```diff
@@ -856,10 +835,10 @@ void nrn_newtonspace_copyto_device(NewtonSpace* ns) {
...
     acc_memcpy_to_device(&(d_ns->perm), &pint, sizeof(int*));
 
-    ppd = (double**)acc_copyin(ns->jacobian, ns->n * sizeof(double*));
+    double** ppd = (double**)acc_copyin(ns->jacobian, ns->n * sizeof(double*));
```
Comment (tristan0x) [inline coreneuron/gpu/nrn_acc_manager.cpp]:
```suggestion
    auto ppd = static_cast<double**>(acc_copyin(ns->jacobian, ns->n * sizeof(double*)));
```

File: coreneuron/gpu/nrn_acc_manager.cpp [hunk 3]
```diff
@@ -886,22 +865,16 @@ void nrn_sparseobj_copyto_device(SparseObj* so) {
...
-    double** ppd;
-
     Elm** d_rowst = (Elm**)acc_copyin(so->rowst, n1 * sizeof(Elm*));
     acc_memcpy_to_device(&(d_so->rowst), &d_rowst, sizeof(Elm**));
 
     Elm** d_diag = (Elm**)acc_copyin(so->diag, n1 * sizeof(Elm*));
     acc_memcpy_to_device(&(d_so->diag), &d_diag, sizeof(Elm**));
 
-    pu = (unsigned*)acc_copyin(so->ngetcall, so->_cntml_padded * sizeof(unsigned));
+    unsigned* pu = (unsigned*)acc_copyin(so->ngetcall, so->_cntml_padded * sizeof(unsigned));
```
Comment (tristan0x) [inline coreneuron/gpu/nrn_acc_manager.cpp]: same here

File: coreneuron/gpu/nrn_acc_manager.cpp [hunk 4]
```diff
@@ -886,22 +865,16 @@ void nrn_sparseobj_copyto_device(SparseObj* so) {
...
     acc_memcpy_to_device(&(d_so->ngetcall), &pu, sizeof(Elm**));
 
-    pd = (double*)acc_copyin(so->rhs, n1 * so->_cntml_padded * sizeof(double));
+    double* pd = (double*)acc_copyin(so->rhs, n1 * so->_cntml_padded * sizeof(double));
```
Comment (tristan0x) [inline coreneuron/gpu/nrn_acc_manager.cpp]: same here

File: coreneuron/gpu/nrn_acc_manager.cpp [hunk 5]
```diff
@@ -940,7 +913,7 @@ void nrn_sparseobj_copyto_device(SparseObj* so) {
     // visit all the Elm again and fill in pelm->r_down and pelm->c_left
     for (unsigned irow = 1; irow < n1; ++irow) {
         for (Elm* elm = so->rowst[irow]; elm; elm = elm->c_right) {
-            pelm = (Elm*)acc_deviceptr(elm);
+            Elm* pelm = (Elm*)acc_deviceptr(elm);
```
Comment (tristan0x) [inline coreneuron/gpu/nrn_acc_manager.cpp]: same here

File: coreneuron/mechanism/register_mech.cpp [hunk 6]
```diff
@@ -429,7 +410,7 @@ void hoc_reg_ba(int mt, mod_f_t f, int type) {
                    corenrn.get_memb_func(mt).sym);
             nrn_exit(1);
     }
-    bam = (BAMech*)emalloc(sizeof(BAMech));
+    auto bam = (BAMech*)emalloc(sizeof(BAMech));
```
Comment (tristan0x) [inline coreneuron/mechanism/register_mech.cpp]: same here

File: coreneuron/mpi/mpispike.cpp [hunk 7]
```diff
@@ -492,11 +487,10 @@ long nrnmpi_long_allreduce(long x, int type) {
 }
 
 void nrnmpi_dbl_allreduce_vec(double* src, double* dest, int cnt, int type) {
-    int i;
     MPI_Op tt;
     assert(src != dest);
     if (nrnmpi_numprocs < 2) {
-        for (i = 0; i < cnt; ++i) {
+        for (int i = 0; i < cnt; ++i) {
```
Comment (tristan0x) [inline coreneuron/mpi/mpispike.cpp]: std::memcpy ?

File: coreneuron/mpi/mpispike.cpp [hunk 8]
```diff
@@ -396,20 +392,19 @@ static void pgvts_op(double* in, double* inout, int* len, MPI_Datatype* dptr) {
...
 
 int nrnmpi_pgvts_least(double* tt, int* op, int* init) {
-    int i;
     double ibuf[4], obuf[4];
     ibuf[0] = *tt;
     ibuf[1] = (double)(*op);
     ibuf[2] = (double)(*init);
     ibuf[3] = (double)nrnmpi_myid;
-    for (i = 0; i < 4; ++i) {
+    for (int i = 0; i < 4; ++i) {
```
Comment (tristan0x) [inline coreneuron/mpi/mpispike.cpp]: std::memcpy ?

File: coreneuron/mpi/mpispike.cpp [hunk 9]
```diff
@@ -513,11 +507,10 @@ void nrnmpi_dbl_allreduce_vec(double* src, double* dest, int cnt, int type) {
 }
 
 void nrnmpi_long_allreduce_vec(long* src, long* dest, int cnt, int type) {
-    int i;
     MPI_Op tt;
     assert(src != dest);
     if (nrnmpi_numprocs < 2) {
-        for (i = 0; i < cnt; ++i) {
+        for (int i = 0; i < cnt; ++i) {
```
Comment (tristan0x) [inline coreneuron/mpi/mpispike.cpp]: std::memcpy ?

File: coreneuron/sim/scopmath/sparse_thread.cpp [hunk 10]
```diff
@@ -202,10 +202,7 @@ int sparse_thread(SparseObj* so,
 int _cvode_sparse_thread(void** v, int n, int* x, SPFUN fun, _threadargsproto_)
 #define x_(arg) _p[x[arg] * _STRIDE]
 {
-    int i, j, ierr;
-    SparseObj* so;
-
-    so = (SparseObj*)(*v);
+    SparseObj* so = (SparseObj*)(*v);
```
Comment (tristan0x) [inline coreneuron/sim/scopmath/sparse_thread.cpp]: same here

```suggestion
    auto so = static_cast<SparseObj*>(*v);
```

File: coreneuron/sim/multicore.cpp [hunk 11]
```diff
@@ -107,7 +107,7 @@ void nrn_threads_create(int n) {
 
 void nrn_threads_free() {
     if (nrn_nthread) {
-        free_memory((void*)nrn_threads);
+        delete[] nrn_threads;
```
Comment (alkino) [inline coreneuron/sim/multicore.cpp]: Not created by me, but it seems to be the reason why the test failed with travis.

**Comment by alkino** [comment]
Please retest

**Comment by alkino** [comment]
Please retest

**Comment by alkino** [comment]
Please test

**Comment by pramodk** [comment]
@alkino : if you are able to build on BB5 with same instructions manually, the only thing could be different is config files. As Jenkins CI nodes are shared, there could be some configs files in `$HOME/.spack` from different job! So [here I just removed](https://github.com/BlueBrain/CoreNeuron/pull/246/commits/e41c1c2b1dcc00f2d53f1eee5b2e66b40fb1b21c) those files and neuron is building fine!

In general I would suggest to change `$HOME` in Jenkinsfile to `$WORKSPACE` so that each CI build has different home directory. Ask @matz-e if have any questions.

Next [compilation errors are from GPU build](https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/oss.coreneuron.pipeline/detail/oss.coreneuron.pipeline/832/pipeline/33). We can use Spack but we should be always able to reproduce this manually. Using [README instructions](https://github.com/BlueBrain/CoreNeuron#building-with-gpu-support), here is what I did : 

```	
kumbhar@bbpv1:~/tmp/xx$ git clone https://github.com/BlueBrain/CoreNeuron.git -b more_clean
Cloning into 'CoreNeuron'...
remote: Enumerating objects: 540, done.
remote: Counting objects: 100% (540/540), done.
remote: Compressing objects: 100% (312/312), done.
remote: Total 11529 (delta 327), reused 333 (delta 225), pack-reused 10989
Receiving objects: 100% (11529/11529), 35.44 MiB | 9.90 MiB/s, done.
Resolving deltas: 100% (8689/8689), done.
kumbhar@bbpv1:~/tmp/xx$ mkdir CoreNeuron/build
kumbhar@bbpv1:~/tmp/xx$ cd CoreNeuron/build/
kumbhar@bbpv1:~/tmp/xx/CoreNeuron/build$ module purge all
kumbhar@bbpv1:~/tmp/xx/CoreNeuron/build$ module load hpe-mpi pgi cmake cmake cuda boost
kumbhar@bbpv1:~/tmp/xx/CoreNeuron/build$ module list
Currently Loaded Modulefiles:
  1) hpe-mpi/2.16          2) pgi/19.4              3) cmake/3.13.0          4) cuda/9.2.88           5) boost/1.68.0/serial

kumbhar@bbpv1:~/tmp/xx/CoreNeuron/build$ cmake ..  -DCMAKE_C_FLAGS:STRING="-DR123_USE_SSE=0 -O2 -Minline=size:1000,levels:100,totalsize:40000,maxsize:4000" -DCMAKE_CXX_FLAGS:STRING="-DR123_USE_SSE=0 -O2 -Minline=size:1000,levels:100,totalsize:40000,maxsize:4000"  -DCOMPILE_LIBRARY_TYPE=STATIC  -DCUDA_HOST_COMPILER=`which gcc` -DCUDA_PROPAGATE_HOST_FLAGS=OFF -DCORENRN_ENABLE_GPU=ON  -DCMAKE_BUILD_TYPE=CUSTOM

-- The C compiler identification is PGI 19.4.0
-- The CXX compiler identification is PGI 19.4.0
-- Check for working C compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/pgi-19.4-hdirysdrvd/linux86-64/19.4/bin/pgcc
-- Check for working C compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/pgi-19.4-hdirysdrvd/linux86-64/19.4/bin/pgcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/pgi-19.4-hdirysdrvd/linux86-64/19.4/bin/pgc++
-- Check for working CXX compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/pgi-19.4-hdirysdrvd/linux86-64/19.4/bin/pgc++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
…

--
-- Configured CoreNEURON
--
-- Some things you can do now:
-- --------------------+--------------------------------------------------------
-- Command             |   Description
-- --------------------+--------------------------------------------------------
-- make install        | Will install CoreNEURON to: /usr/local
-- --------------------+--------------------------------------------------------
--  Build option       | Status
-- --------------------+--------------------------------------------------------
-- MPI                 | ON
--   INC               | /opt/hpe/hpc/mpt/mpt-2.16/include
-- OpenMP              | ON
-- NMODL               | OFF
-- MOD2C               | MOD2C_PROJ-NOTFOUND
-- GPU Support         | ON
-- CUDA Unified Memory | OFF
-- Auto Timeout        | ON
-- Wrap exp()          | OFF
-- SplayTree Queue     | ON
-- NetReceive Buffer   | ON
-- Caliper             | OFF
-- Likwid              | OFF
-- Unit Tests          | ON
-- ReportingLib        | OFF
--   SONATA support    | OFF
-- --------------+--------------------------------------------------------------
--  See documentation : https://github.com/BlueBrain/CoreNeuron/
-- --------------+--------------------------------------------------------------
--
-- Configuring done
-- Generating done
-- Build files have been written to: /gpfs/bbp.cscs.ch/home/kumbhar/tmp/xx/CoreNeuron/build
```

and building should reproduce the error:


```
kumbhar@bbpv1:~/tmp/xx/CoreNeuron/build$ make -j
Scanning dependencies of target mod2c
Scanning dependencies of target scopmath
[  1%] Building NVCC (Device) object coreneuron/CMakeFiles/cudacoreneuron.dir/utils/randoms/cudacoreneuron_generated_nrnran123.cu.o
[  2%] Building NVCC (Device) object coreneuron/CMakeFiles/cudacoreneuron.dir/utils/profile/cudacoreneuron_generated_cuda_profile.cu.o
Scanning dependencies of target swap_endian_nounroll.cpp_test_bin
Scanning dependencies of target swap_endian_default.cpp_test_bin
Scanning dependencies of target swap_endian_oddunroll.cpp_test_bin
Scanning dependencies of target swap_endian_noasm.cpp_test_bin
Scanning dependencies of target endianness_test.cpp_test_bin
Scanning dependencies of target alignment_test_bin
[  3%] Creating directories for 'mod2c'
[  4%] Building CXX object coreneuron/CMakeFiles/scopmath.dir/sim/scopmath/abort.cpp.o
[  4%] Building CXX object coreneuron/CMakeFiles/scopmath.dir/sim/scopmath/ssimplic_thread.cpp.o

….
"/gpfs/bbp.cscs.ch/home/kumbhar/tmp/xx/CoreNeuron/coreneuron/mechanism/capac.cp
          p", line 157: internal error: assertion failed:
          pop_namespace_reactiveation_scope: entry not reactivation extension
          (../src/scope_stk.c, line 9775)

  }
```

If you diff the changes of  mechanism/capac.cpp with master branch you will see what is causing the error?

File: coreneuron/mechanism/capac.cpp [hunk 12]
```diff
@@ -65,87 +65,72 @@ void nrn_mul_capacity(NrnThread*, Memb_list*, int);
...
     register_mech(mechanism, nrn_alloc_capacitance, (mod_f_t)0, (mod_f_t)0, (mod_f_t)0,
                   (mod_f_t)nrn_init_capacitance, -1, 1);
-    mechtype = nrn_get_mechtype(mechanism[1]);
+    int mechtype = nrn_get_mechtype(mechanism[1]);
     _nrn_layout_reg(mechtype, LAYOUT);
     hoc_register_prop_size(mechtype, nparm, 0);
 }
 
-#define cm vdata[0 * _STRIDE]
-#define i_cap vdata[1 * _STRIDE]
```
Comment (pramodk) [inline coreneuron/mechanism/capac.cpp]: I understand why we should remove this but at the same time it helps lot in reading below code :

```
- i_cap = cfac * cm * _vec_rhs[ni[_iml]];
+ vdata[1 * _STRIDE] = cfac * vdata[0 * _STRIDE] * _vec_rhs[ni[_iml]];
```

File: coreneuron/io/mk_mech.cpp [hunk 13]
```diff
@@ -101,8 +101,7 @@ void mk_mech(const char* datpath) {
     fs.close();
 
     fname = sdprintf(fnamebuf, sizeof(fnamebuf), "%s/%s", datpath, "byteswap1.dat");
-    FILE* f;
-    f = fopen(fname, "r");
```
Comment (matz-e) [inline coreneuron/io/mk_mech.cpp]: `fstream`?

File: coreneuron/mechanism/capac.cpp (hunk 12 above)
Comment (alkino) [reply coreneuron/mechanism/capac.cpp]: ok, I can parametrized macro and capitalize them.

Should be more readable.

File: coreneuron/io/mk_mech.cpp (hunk 13 above)
Comment (alkino) [reply coreneuron/io/mk_mech.cpp]: of course.

File: coreneuron/sim/finitialize.cpp [hunk 14]
```diff
@@ -49,42 +46,41 @@ void nrn_finitialize(int setv, double v) {
...
+            for (int i = 0; i < _nt->end; ++i) {
                 vec_v[i] = v;
             }
         }
     }
 
     if (nrn_have_gaps) {
         nrnmpi_v_transfer();
-        for (i = 0; i < nrn_nthread; ++i) {
+        for (int i = 0; i < nrn_nthread; ++i) {
```
Comment (ohm314) [inline coreneuron/sim/finitialize.cpp]: maybe better to iterate directly over the pointer

File: coreneuron/sim/scopmath/ssimplic_thread.cpp [hunk 15]
```diff
@@ -47,29 +45,24 @@ int _ss_sparse_thread(SparseObj* v,
...
+    bool flag = true;
+    for (int i = 0; i < n; i++) {
         if (s_(i) < -1e-6) {
             s_(i) = 0.;
-            flag = 0;
+            flag = false;
         }
     }
-    return flag;
+    return flag ? 1 : 0;
```
Comment (ohm314) [inline coreneuron/sim/scopmath/ssimplic_thread.cpp]: can this function not simply return bool?

**Comment by alkino** [comment]
Please retest

File: coreneuron/mechanism/capac.cpp (hunk 12 above)
Comment (pramodk) [reply coreneuron/mechanism/capac.cpp]: This is derived from mod2c and I wouldn't bother with these auto generated files atm. With NMODL they will get replaced.

File: coreneuron/mechanism/patternstim.cpp [hunk 16]
```diff
@@ -199,7 +201,7 @@ Point_process* nrn_artcell_instantiate(const char* mechname) {
     int psize = corenrn.get_prop_param_size()[type];
     int dsize = corenrn.get_prop_dparam_size()[type];
     // int layout = nrn_mech_data_layout[type]; // not needed because singleton
-    Memb_list* ml = tml->ml;
+    Memb_list* ml = nt->tml->ml;
```
Comment (pramodk) [inline coreneuron/mechanism/patternstim.cpp]: Just for the record : this and previous change is not doing same thing as old code (`nt->tml->ml` is always first element here).

Comment (ohm314) [reply coreneuron/mechanism/patternstim.cpp]: Uff, good catch!

File: coreneuron/mechanism/capac.cpp (hunk 12 above)
Comment (ohm314) [reply coreneuron/mechanism/capac.cpp]: It was not clear that we want in the future to generate those files again. In this case I would not touch them at all.

**Comment by pramodk** [comment]
@alkino : finally all green! good work!

**Comment by alkino** [comment]
This patch will not contains new clean after it will be hard to merge / rebase with other current development.

See #250 for what there is still to do.

So I will not change what you request if it's not something already contains in this patch.

**Comment by alkino** [comment]
Everything is fixed, let's merge once it's green!

**Comment by pramodk** [comment]
Please retest

File: coreneuron/mechanism/patternstim.cpp [hunk 17]
```diff
@@ -160,6 +160,33 @@ size_t read_raster_file(const char* fname, double** tvec, int** gidvec) {
...
+    tml->ml->data = (double*)ecalloc(tml->ml->nodecount * psize, sizeof(double));
+    tml->ml->pdata = (Datum*)ecalloc(tml->ml->nodecount * dsize, sizeof(Datum));
+    tml->ml->_thread = nullptr;
+    tml->ml->_net_receive_buffer = nullptr;
+    tml->ml->_net_send_buffer = nullptr;
+    tml->ml->_permute = nullptr;
+
+    return tml;
+}
+
```
Comment (pramodk) [inline coreneuron/mechanism/patternstim.cpp]: this is more clear! thanks!

**Comment by pramodk** [comment]
@alkino : done! great! thanks!

File: coreneuron/io/nrn_checkpoint.cpp [hunk 18]
```diff
@@ -162,16 +162,15 @@ void write_checkpoint(NrnThread* nt, int nb_threads, const char* dir, bool swap_
...
     swap_bytes = swap_bytes_order;
 
     /**
      * if openmp threading needed:
      *  #pragma omp parallel for private(i) shared(nt, nb_threads) schedule(runtime)
      */
     FileHandlerWrap f;
-    for (i = 0; i < nb_threads; i++) {
-        if (nt[i].ncell || nt[i].tml) {
+    for (int i = 0; i < nb_threads; i++) {
```
[the remark this answers is not in the document]
Comment (alkino) [reply coreneuron/io/nrn_checkpoint.cpp]: Unfortunately not...

`NrnThread` is a c-style array, not linked and not null terminated... We have to index it.

**Pull request merged**